ufl/constantvalue.py

```diff
@@ -249,12 +254,63 @@ def __float__(self):
     def __int__(self):
         return int(self._value)
 
+    def __complex__(self):
+        return complex(self._value)
+
     def __neg__(self):
         return type(self)(-self._value)
 
     def __abs__(self):
         return type(self)(abs(self._value))
 
+    def real(self):
+        return self._value.real
+
+    def imag(self):
+        return self._value.imag
+
+@ufl_type(wraps_type=complex, is_literal=True)
+class ComplexValue(ScalarValue):
+    "UFL literal type: Representation of a constant, complex scalar"
+    __slots__= ()
+
+    def __getnewargs__(self):
+        return (self._value,)
+
+    def __new__(cls, value):
+        if value.imag == 0:
+            if value.real == 0:
+                return Zero()
+            else:
+                return FloatValue.__new__(cls,value.real)
+        else:
+            return ConstantValue.__new__(cls,value)
+
+    def __init__(self,value):
+        ScalarValue.__init__(self, complex(value))
+
+    def modulus(self):
+        return sqrt(self.value().real**2 + self.value().imag**2)
```

Comment:
Surely this exists already for complex numbers in Python, so we don't need to spell out what the magnitude is...